### ISSUE TITLE
Queries for pi.hole being forwarded to upstream

### ISSUE THREAD

**Issue by tgupta3** [pi-hole/docker-pi-hole#600]
I am continuously observing the following set of lines where the query is being forwarded to the upstream DNS Server
```
Apr 18 18:43:20 dnsmasq[507]: query[A] pi.hole from 127.0.0.1
Apr 18 18:43:20 dnsmasq[507]: forwarded pi.hole to 54.183.15.10
Apr 18 18:43:20 dnsmasq[507]: nameserver 54.183.15.10 refused to do a recursive query
```
Tcpdump 
```
18:44:53.076076 IP 10.100.0.100.13097 > 54.183.15.10.53: 62636 [1au] A? pi.hole. (36)
18:44:53.092186 IP 54.183.15.10.53 > 10.100.0.100.13097: 62636- 0/0/1 (36)
```

This is a... 

- [ ] Request for a new or modified feature
- [x ] Issue trying to run the docker image
- [ ] Issue trying to build / test / develop the docker image

## Description

Observing queries to pi.hole being forwarded to upstream DNS. However if a client queries the pi.hole server with a request for pi.hole, it's correctly responded

From the docker host
```
dig pi.hole @10.100.0.100 +short
127.0.0.1
10.100.0.100
```
Using docker exec queries are responded correctly
```
~/etc/pihole-100# docker exec -it pihole_100 dig @172.18.0.2 pi.hole +short
127.0.0.1
10.100.0.100

~/etc/pihole-100# docker exec -it pihole_100 dig @127.0.0.1 pi.hole +short
127.0.0.1
```
Relevant Logs
```
Apr 18 18:49:03 dnsmasq[507]: /etc/hosts pi.hole is 127.0.0.1
Apr 18 18:49:06 dnsmasq[507]: query[A] pi.hole from 172.18.0.2
Apr 18 18:49:06 dnsmasq[507]: /etc/hosts pi.hole is 127.0.0.1
Apr 18 18:49:06 dnsmasq[507]: /etc/pihole/local.list pi.hole is 10.100.0.100
Apr 18 18:49:12 dnsmasq[507]: query[A] pi.hole from 127.0.0.1
Apr 18 18:49:12 dnsmasq[507]: /etc/hosts pi.hole is 127.0.0.1
Apr 18 18:49:14 dnsmasq[507]: query[A] pi.hole from 127.0.0.1
Apr 18 18:49:14 dnsmasq[507]: /etc/hosts pi.hole is 127.0.0.1
```
## Expected Behavior

Query should not get forwarded to upstream DNS
## Actual Behavior

Query getting forwarded to upstream DNS
## Steps to Reproduce and debugging done

e.g. your docker run command, pages to visit, CLI commands you ran
1. Running this via ansible if this matters. /etc/hosts has 127.0.0.1
2. 
3. 
4. 

## Debug steps I have tried

- [x ] I have tried destroying my container instance, pulling the newest image version, and re-creating a new container
- [ x] I have tried running the nearly stock `docker run` example in the readme (removing any customizations I added)
- [x ] I have tried running without my volume data mounts to eliminate volumes as the cause
- [x ] I have searched this repository for existing issues and pull requests that look similar 

## Your Environment 

* Docker Host Operating System and OS Version: 
* Docker Version:
* Hardware architecture:

**Comment by tgupta3** [comment]
This seems to be caused by healthcheck
```
root@grasshopper:~/etc/pihole-100# dig +norecurse pi.hole @10.100.0.100 

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> +norecurse pi.hole @10.100.0.100
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30435
;; flags: qr ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 2800
;; QUESTION SECTION:
;pi.hole.			IN	A

;; Query time: 17 msec
;; SERVER: 10.100.0.100#53(10.100.0.100)
;; WHEN: Sat Apr 18 19:11:59 PDT 2020
;; MSG SIZE  rcvd: 36
```
```
Apr 18 19:11:59 dnsmasq[500]: query[A] pi.hole from 10.100.0.100
Apr 18 19:11:59 dnsmasq[500]: forwarded pi.hole to 54.183.15.10
Apr 18 19:11:59 dnsmasq[500]: nameserver 54.183.15.10 refused to do a recursive query

```

**Comment by dever23b** [comment]
I've been trying to fight this issue for the past couple of weeks.  I don't know if it's always been around or if something in my config causes it, but it's occurring on all of my pihole containers.  I don't know if this helps at all, but I've narrowed the impactful difference down to the +norecurse option in dig.  If you use nslookup or run dig with out +norecurse, the response comes back fine (provided there's not some kinda hiccup in your local.list file).  But, with +norecurse, pihole forwards the request upstream and it fails and then triggers the healthcheck to report pihole as unhealthy.

Looking forward to a solution on this.

**Comment by tgupta3** [comment]
Yes it was the no recurse option that was causing this. I ma using ansible to deploy this, so just change the health check to remove no-recurse and also specifying the server ip so as to emulate a real DNS query. The workaround seems to work, as the queries are being answered based on local.list now. 
```
healthcheck:
      test: "dig +retry=0 @{{ docker_pihole_100.server_ip }} pi.hole || exit 1"
```

**Comment by dever23b** [comment]
That definitely works.  What I can't figure out is why +norecurse fails.  

#410 has a comment stating that pihole's failure to answer that query without recursion would indicate a problem...

**Comment by tgupta3** [comment]
I think this answers it
https://bugzilla.redhat.com/show_bug.cgi?id=1755610#c3

 If i can recall in recursive option, queries can be forwarded to upstream DNS or one is giving the right to the DNS resolver to query other resolver on behalf of the client. In iterative, it's like a boolean you either have an answer or you don't.

With no[recurse] option set, query should not be forwarded upstream i.e it's an iterative query. So you either answer it correctly or you don't which doesn't seem to match pi-hole behavior.

**Comment by lightswitch05** [comment]
> So you either answer it correctly or you don't

Why would your pihole not know the answer to a `pi.hole` query?

**Comment by dever23b** [comment]
Unknown.  I'm not an expert in DNS and don't fully understand what's happening, but my experience has been this: a normal query using dig/nslookup/systemd-resolve/etc triggers the expected response: an IP of the pihole server responding.  pihole.log also shows that the query was answered from local.list.  Based on my (admittedly, very limited) understanding of norecurse, it seems like I should expect the opposite behavior to what actually happens: with norecurse, I'd expect the query to -not- get forwarded and instead only get a local answer from the selected DNS server.  But what happens is that without norecurse, dnsmasq answers locally via the entry in local.list and, with norecurse, it forwards the query upstream (what I'd consider to be "recursion") in a failure.

For quite some time, I couldn't manually reproduce the issue: every request I made from every host worked as expected, yet I was seeing log entries very often indicating requests for pi.hole being forwarded (not my test requests, though).  I finally discovered it was the healthcheck and specifically the use of the +norecurse option.  When dig +norecurse is done, for whatever reason, pihole skips the local.list and forwards the request upstream.  I have not been able to discover any configuration of dnsmasq that causes it to trap the norecurse request and answer it.  I toyed around with address=, local=, and server=, to no avail: the request always got forwarded.  I can't remember for sure, but I believe I also tried a server= to direct the 'pi.hole' requests to another local DNS server too.

**Comment by edgd1er** [comment]
Well, I noticed the same as you guys, the dig norecurse seems to do the opposite as expected. For the test, I changed the container port to isolate it from external input requests, and set DEBUG_AL=true. I'm running docker from a debian 9.12

TLDR: in docker-compose or docker command line,  redefine HEALTHCHECK cmd without the +norecurse.

When I run these three commands, answer is correct (127.0.0.1) and request is not forwarded.
* from host: `dig @192.168.53.210 -p 1053 pi.hole`
* from host: `dig @127.0.0.1 -p 1053 pi.hole`
* from container: `dig @127.0.0.1 pi.hole`
``` 
[2020-04-27 15:19:06.429 544] Waiting for lock in _FTL_new_query() (dnsmasq_interface.c:38)
[2020-04-27 15:19:06.429 544] Obtained lock for _FTL_new_query() (dnsmasq_interface.c:38)
[2020-04-27 15:19:06.429 544] Removed lock in _FTL_new_query() (dnsmasq_interface.c:92)
[2020-04-27 15:19:06.429 544] Waiting for lock in _FTL_cache() (dnsmasq_interface.c:661)
[2020-04-27 15:19:06.429 544] Obtained lock for _FTL_cache() (dnsmasq_interface.c:661)
[2020-04-27 15:19:06.429 544] Removed lock in _FTL_cache() (dnsmasq_interface.c:677)
[2020-04-27 15:19:06.429 544] Waiting for lock in _FTL_cache() (dnsmasq_interface.c:661)
[2020-04-27 15:19:06.429 544] Obtained lock for _FTL_cache() (dnsmasq_interface.c:661)
[2020-04-27 15:19:06.429 544] Removed lock in _FTL_cache() (dnsmasq_interface.c:677)
```

When I add the +norecurse parameter, request failed, after being forwarded to external dns.
* from host: `dig +short +norecurse @192.168.53.210 -p 1053 pi.hole`
* from host: `dig +short +norecurse @127.0.0.1 -p 1053 pi.hole`
* from container: `dig +short +norecurse @127.0.0.1 pi.hole`
``` 
[2020-04-27 15:23:39.527 544] Waiting for lock in _FTL_new_query() (dnsmasq_interface.c:38)
[2020-04-27 15:23:39.527 544] Obtained lock for _FTL_new_query() (dnsmasq_interface.c:38)
[2020-04-27 15:23:39.527 544] Removed lock in _FTL_new_query() (dnsmasq_interface.c:92)
[2020-04-27 15:23:39.527 544] Waiting for lock in _FTL_forwarded() (dnsmasq_interface.c:257)
[2020-04-27 15:23:39.527 544] Obtained lock for _FTL_forwarded() (dnsmasq_interface.c:257)
[2020-04-27 15:23:39.527 544] **** forwarded pi.hole to 80.67.169.12 (ID 72, dnsmasq/forward.c:558)
[2020-04-27 15:23:39.527 544] Removed lock in _FTL_forwarded() (dnsmasq_interface.c:279)
[2020-04-27 15:23:39.527 544] Waiting for lock in _FTL_forwarded() (dnsmasq_interface.c:257)
[2020-04-27 15:23:39.528 544] Obtained lock for _FTL_forwarded() (dnsmasq_interface.c:257)
[2020-04-27 15:23:39.528 544] **** forwarded pi.hole to 80.67.169.40 (ID 72, dnsmasq/forward.c:558)
[2020-04-27 15:23:39.528 544] Removed lock in _FTL_forwarded() (dnsmasq_interface.c:279)
```

the beginning of processing is the same but after dnsmasq_interface.c L92, the request is handled in a different way (recurse =>dnsmasq_interface.c:661, norecurse=>dnsmasq_interface.c:257). That might by an hint for someone who whish to go further in analysis. this is where I stopped.

In the docker-compose.yml, i redefined the HEALTHCHECK CMD without the `+norecurse`.

**Comment by dever23b** [comment]
Yeah, we figured out the healthcheck workaround above.  Thanks for the trace.  I'm pretty rusty on my C, but I went digging through the source starting from those references to dnsmasq_interface.c you found.  Again, my C is rusty, so maybe I'm misinterpreting it.  

Based on your trace and your observation that the request is handled differently, I tried to figure out where that logic occurs.  Two areas of code that caught my eye.  I backtraced the call to dnsmas_interface.c:661 to forward.c:1641. 

forward.c:receive_query() calls answer_request() and, if answer_request() returns a non-zero response, attempts to resolve the query locally; else, it forwards the query.
```
      m = answer_request(header, ((char *) header) + udp_size, (size_t)n,
			 dst_addr_4, netmask, now, ad_reqd, do_bit, have_pseudoheader);

      if (m >= 1)
	{
	  send_from(listen->fd, option_bool(OPT_NOWILD) || option_bool(OPT_CLEVERBIND),
		    (char *)header, m, &source_addr, &dst_addr, if_index);
	  daemon->metrics[METRIC_DNS_LOCAL_ANSWERED]++;
	}
      else if (forward_query(listen->fd, &source_addr, &dst_addr, if_index,
			     header, (size_t)n, now, NULL, ad_reqd, do_bit))
	daemon->metrics[METRIC_DNS_QUERIES_FORWARDED]++;
      else
	daemon->metrics[METRIC_DNS_LOCAL_ANSWERED]++;
    }
```

rfc1035.c: answer_request() returns 0 if a query cannot be answered from cache or alternatively a packet size.  One of the first checks answer_request() does it to see whether recursion was enabled on the query and, if not, it immediately exits 0.
```
  /* never answer queries with RD unset, to avoid cache snooping. */
  if (!(header->hb3 & HB3_RD) ||
      ntohs(header->ancount) != 0 ||
      ntohs(header->nscount) != 0 ||
      ntohs(header->qdcount) == 0 ||
      OPCODE(header) != QUERY )
    return 0;
```

I believe what may be happening is that, when norecurse is set, FTL doesn't even attempt to answer the query locally and instead forwards it.  Again, I'm not an expert on DNS protocol, but this logic is opposite to what I'd expect.  It seems like, if norecurse is set, FTL should only attempt to answer the query locally; if recursion is requested, then FTL would attempt locally first before forwarding the query. 

With all that said, I do not believe this bug is in the correct repo: I believe the healthcheck in the Dockerfile is correct, and the bug is actually in the FTL behavior.  I'm going to search for a bug report in the FTL repo and, if necessary, create a new one referencing this.

**Comment by dever23b** [comment]
Issue in pi-hole/FTL: https://github.com/pi-hole/FTL/issues/738

**Comment by dever23b** [comment]
Pi-Hole FTL v5.0 was recently released.  After upgrading, I'm not seeing this issue anymore.  I'm not sure if it's just me (@DL6ER originally reported that he was unable to reproduce this directly, while still confirming the bug existed).  It may be worth your time to try upgrading Pi-Hole and seeing if you can still reproduce the issue.

**Comment by tgupta3** [comment]
I have already upgraded but still have the workaround in place. I'll try to see if removing it fixes it.
